### ISSUE TITLE
shiny app does not connect when `uiPattern` is long path.

### ISSUE THREAD

**Issue by jtlandis** [rstudio/shiny#3623]
I am trying to utilize `shinyApp(uiPattern = ".*")` so that I could run an app and then grab the session$request$PATH_INFO later. In my testing, I am seeing some unexpected results whenever the path has more than a single delimiter.

Below is a minimal reproducible example of the issue.

```R
library(shiny)
ui <- fluidPage(
  wellPanel(
    textInput("text", "Input"),
    verbatimTextOutput("out")
  )
)

server <- function(input, output, session) {
  #for debugging
  print(session$request$PATH_INFO)
  output$out <- renderPrint({
    input$text
  })
}

app <- shinyApp(ui, server, uiPattern = ".*")

runApp(app, port = 1234, launch.browser = F)
```

If you point your browser to `http://127.0.0.1:1234/`, you will see the app as intended as well if you point to `http://127.0.0.1:1234/foo/`

If you try to make a path like `http://127.0.0.1:1234/foo/bar`, then the html is rendered improperly and the shiny server session is never initiated. 

Does anyone know what the issue is and if it is fixable?

I would assume that `uiPattern = ".*"` would sufficiently capture a path like `/foo/bar`. I assume it is being handled properly because the html is rendered and the status by the handler is 200, which makes me believe the error is somewhere else.

**Comment by jtlandis** [comment]
I was performing some debugging (by placing a message in line 233 of shiny's shinyui.R file) and it seems that in the browser there are several calls to other .js and .css files upon viewing a page with a long path, whereas on a short path, `uiHttpHandler()` (the function in which my debugging message is) is only called once.

**Comment by jtlandis** [comment]
I did some debugging using my chrome browser, and I think I am closer to understanding the issue.

Whenever a request is made to the browser, the first response returns the proper html regardless of the domain path 

<details>

    <!DOCTYPE html>
    <html>
    <head>
      <meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
      <script type="application/shiny-singletons"></script>
      <script type="application/html-dependencies">jquery[3.6.0];shiny-css[1.7.1.9003];shiny-javascript[1.7.1.9003];bootstrap[3.4.1]</script>
    <script src="jquery-3.6.0/jquery.min.js"></script>
    <link href="shiny-css-1.7.1.9003/shiny.min.css" rel="stylesheet" />
    <script src="shiny-javascript-1.7.1.9003/shiny.min.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <link href="bootstrap-3.4.1/css/bootstrap.min.css" rel="stylesheet" />
    <link href="bootstrap-3.4.1/accessibility/css/bootstrap-accessibility.min.css" rel="stylesheet" />
    <script src="bootstrap-3.4.1/js/bootstrap.min.js"></script>
    <script src="bootstrap-3.4.1/accessibility/js/bootstrap-accessibility.min.js"></script>
    </head>
    <body>
      <div class="container-fluid">
        <div class="well">
          <div class="form-group shiny-input-container">
            <label class="control-label" id="text-label" for="text">Input</label>
            <input id="text" type="text" class="form-control" value=""/>
          </div>
          <pre class="shiny-text-output noplaceholder" id="out"></pre>
        </div>
      </div>
    </body>
    </html>

</details>

After a request to say `/foo/bar`

it makes the following request for the resources named in the header... namely

* jquery-3.6.0/jquery.min.js
* shiny-css-1.7.1.9003/shiny.min.css
* bootstrap-3.4.1/css/bootstrap.min.css
* shiny-javascript-1.7.1.9003/shiny.min.js
* bootstrap-3.4.1/accessibility/css/bootstrap-accessibility.min.css
* bootstrap-3.4.1/js/bootstrap.min.js
* bootstrap-3.4.1/accessibility/js/bootstrap-accessibility.min.js

However, since the html file that references these resources is being rendered to the path `http://127.0.0.1:1234/foo/bar`, it looks for these resources in `http://127.0.0.1:1234:/foo` instead of where they can really be found `http://127.0.0.1:1234`

performing a curl while the server is running verifies this result. Any request to `http://127.0.0.1:1234/foo/*` returns the same html file listed in details, whereas a request to `http://127.0.0.1:1234/jquery-3.6.0/jquery.min.js` returns the proper jQuery.min.js file.

**Comment by jcheng5** [comment]
Yes, you nailed it with your last comment. In your original comment you said http://127.0.0.1:1234/foo/ worked, I wouldn’t expect this to be the case. Unless you actually meant http://127.0.0.1:1234/foo.

**Comment by jtlandis** [comment]
This may be an issue for `htmltools`. If you add the following after the 12th line in `renderDocument()` function

    deps <- lapply(deps, function(x){x$src$href <- paste0("/", x$src$href); x})
   
This forces all the href paths to be absolute instead of relative and the page loads the `js` dependencies properly.

**Comment by jtlandis** [comment]
My specific use case was initially intended to be a hacky solution to nginx header passing to a shiny App I was hosting. My idea was to pass the full path to the app so that I could store it as a redirect later.

However, I could see potential use cases for allowing uiPattern to work with long queries. This could allow for some dynamic rendering of web pages when using a function for the `ui` argument of `shinyApp`.

**Comment by jtlandis** [comment]
Pardon my ignorance, but do you have an example of when a shiny app is accessed by an arbitrary path (that does not use a reverse proxy)? Is this specific to shiny server, or something else?

**Comment by jcheng5** [comment]
> My specific use case was initially intended to be a hacky solution to nginx header passing to a shiny App I was hosting. My idea was to pass the full path to the app so that I could store it as a redirect later.

Would a URL query argument be better for that? Those are pretty easily accessible in Shiny, from both UI function and server function.

> Pardon my ignorance, but do you have an example of when a shiny app is accessed by an arbitrary path (that does not use a reverse proxy)? Is this specific to shiny server, or something else?

I meant using a reverse proxy (which to me includes Shiny Server, Connect, Nginx, Apache...).

**Comment by drghirlanda** [comment]
I think this should still be considered a bug, because once you use uiPattern you cannot control what URL the user asks for. Maybe you don't have / want / need to handle long paths, but a request of that kind can still be made to the shiny app, right? There should be at least a mechanism to intercept these cases and flag them as invalid to the user.

**Comment by jtlandis** [comment]
It has been a while since I opened this issue, but I have found a workaround in my own testing for anyone who may stumble onto this issue and want to handle long paths appropriately. 

The jist is, you _can handle any long path_, the trick is how do you get all the other html dependencies to be handled correctly (for example .css and .js files).

In order to do this, you must make your `ui` object a function that accepts a `req` ROOK environment and handles the response. This is a lesser known feature in shiny, but here is an [example of how you might do it](https://gist.github.com/jcheng5/2aaff19e67079840350d08361fe7fb20) (it is a guide for handling a "POST" request, but for our purpose, we only need it to handle "GET" requests). 

```
ui <- function(req) {
...
}
```

My `ui` handler function does the following:

1. Check if `req$PATH_INFO` matches any of the expected paths for my app, if so, handle the response by returning the approprate `httpResponse` (you may need to use `shiny:::renderPage()` to print the appropriate content in the response).
2. Otherwise, if the path matches no locations AND `"HTTP_REFERER"` is in the names of the `req` object, then I can assume that this path is a requesting a dependency, so I modify the `PATH_INFO` appropriately and send a `httpResponse(status = 301, headers = list(Location = sprintf("https://%s/app_root/%s", host, modified_path))`

I believe this is the basic summary of how to get it going and it works well enough, but very hacky. Hope this helps.

**Comment by drghirlanda** [comment]
> It has been a while since I opened this issue, but I have found a workaround in my own testing for anyone who may stumble onto this issue and want to handle long paths appropriately.

This is extremely helpful and even cleaner to implement a full website in shiny! Thanks a lot! (I still think the original issue should qualify as a bug as it leads to unwanted behavior to user input that you cannot control or validate.)
